### PR TITLE
Interface Revamp for Numerics Testing

### PR DESCRIPTION
**Pull request by blallen**
### Description

This PR accomplishes two things:

1) finally separates the compressible navier stokes models from their ocean dependencies (accomplished)
2) introduces a new interface to make our lives easier (forthcoming, needed one to ease debugging)

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -39,3 +39,6 @@ src/**/Manifest.toml
 
 # Julia System Images
 *.so
+
+# Julia Files
+*debug_zone*
```
Comment (sandreza) [inline .gitignore:44]: We will want to remove this

File: test/Numerics/DGMethods/advection_diffusion/advection_diffusion_model_1dimex_bjfnks.jl [hunk 2]
```diff
@@ -309,10 +309,10 @@ let
     numlevels = integration_testing ? 4 : 1
 
     @testset "$(@__FILE__)" begin
-        for FT in (Float64, Float32)
-            result = zeros(FT, numlevels)
-            for dim in 2:3
-                for fluxBC in (true, false)
+        for FT in [Float64]
```
Comment (sandreza) [inline test/Numerics/DGMethods/advection_diffusion/advection_diffusion_model_1dimex_bjfnks.jl:312]: We should revert the changes back here

**Review by bischtob — APPROVED** [review]
LGTM!

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl [hunk 3]
```diff
@@ -471,31 +431,74 @@ function numerical_flux_first_order!(
...
+        physics.buoyancy,
+        bcs,
+        ρₒ = params.ρₒ,
+        cₛ = params.cₛ,
+    )
+
+    numerical_flux_first_order = model.numerics.flux # should be a function
+
+    rhs = DGModel(
+        balance_law,
```
Comment (jiahe23) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:496]: Why do we need to reconstruct a balance_law inside this DGModel constructor while the SpatialModel contains a balance_law?

Can we directly use `model.balance_law` here?

Comment (sandreza) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:496]: Things could definitely be clearer. SpatialModel does not contain a balance law, but rather the tag for dispatch to construct a balance law

Comment (blallen) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:504]: Julia is weird that way

**Comment by blallen** [comment]
bors try

File: .gitignore (hunk 1 above)
Comment (blallen) [reply .gitignore]: done

File: test/Numerics/DGMethods/advection_diffusion/advection_diffusion_model_1dimex_bjfnks.jl (hunk 2 above)
Comment (blallen) [reply test/Numerics/DGMethods/advection_diffusion/advection_diffusion_model_1dimex_bjfnks.jl]: also done

**Comment by blallen** [comment]
bors try

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    end
+
+    periodicity = periodicityof(Ω)
+
+    topl = topology(
+        mpicomm,
+        brickrange;
+        periodicity = periodicity,
+        boundary = boundary,
+        connectivity = :face,
```
Comment (blallen) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl:139]: does anyone know why this causes things to segfault? @jkozdon @skandalaCLIMA @simonbyrne 

this change is needed or else the 2D tests fail (seems to make sense based on the other changes from #2071 ) but the 3D test also segfaulted without this (e.g. with `connectivity = :full`).

All of this was working perfectly fine before rebasing on top of master

Comment (skandalaCLIMA) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl:139]: If we have both 2D and 3D examples, I would recommend using
connectivity = dimension == 3 ? :full :  :face

Comment (blallen) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl:139]: can this be done automatically in the constructor? seems like a lot of work for the user to remember if you can't actually have the different combinations

Comment (skandalaCLIMA) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/shared_source/grids.jl:139]: Yes, we eventually plan to support full connectivity for all cases. We will make sure to set the defaults once we add the support.

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
bors try

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl [hunk 5]
```diff
@@ -471,31 +466,83 @@ function numerical_flux_first_order!(
...
 end
 
 include("bc_momentum.jl")
-include("bc_tracer.jl")
+include("bc_temperature.jl")
+
+"""
+STUFF FOR ANDRE'S WRAPPERS
+"""
+
```
Comment (charleskawczynski) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:497]:
```suggestion
```
?

File: test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl [hunk 6]
```diff
@@ -471,31 +466,83 @@ function numerical_flux_first_order!(
...
+    balance_law = CNSE3D{FT}(
+        initial_conditions,
+        (Lˣ, Lʸ, Lᶻ),
+        physics.advection,
+        physics.dissipation,
+        physics.coriolis,
+        physics.buoyancy,
+        bcs,
+        ρₒ = params.ρₒ,
+        cₛ = params.cₛ,
```
Comment (charleskawczynski) [inline test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:534]: `cₛ` does not render in the browser, nor my editor. Can we change the name?

Comment (blallen) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:534]: Screen Shot 2021-03-04 at 17 56 46
it displays fine for me?

Comment (charleskawczynski) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:534]: This is what I see:

Screen Shot 2021-03-05 at 8 58 29 AM

Also, if I copy-paste this into sublime:

Screen Shot 2021-03-05 at 8 59 42 AM

Comment (charleskawczynski) [reply test/Numerics/DGMethods/compressible_navier_stokes_equations/three_dimensional/ThreeDimensionalCompressibleNavierStokesEquations.jl:534]: @blallen ?
